File: RootTools2/src/main/java/com/rarnu/tools/neo/utils/ComponentUtils.java
package com.rarnu.tools.neo.utils;

import android.content.Context;
import android.content.pm.PackageManager;
import com.rarnu.tools.neo.data.CompInfo;
import com.rarnu.tools.neo.root.CommandResult;
import com.rarnu.tools.neo.root.RootUtils;

import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("Duplicates")
public class ComponentUtils {

    public static boolean applicationFreeze(String pkgName, boolean isDisable) {
        String cmd = String.format("pm %s %s", (isDisable ? "disable" : "enable"), pkgName);
        CommandResult result = RootUtils.runCommand(cmd, true);
        return result.error.equals("");
    }

    public static boolean componentFreeze(String pkgName, String componentName, boolean isDisable) {
        String cmd = String.format("pm %s %s/%s", (isDisable ? "disable" : "enable"), pkgName, componentName);
        CommandResult result = RootUtils.runCommand(cmd, true);
        return result.error.equals("");
    }

    public static List<CompInfo> getActivityList(Context ctx, Object pkg) {
        List<CompInfo> lstComponentInfo = new ArrayList<>();
        if (ctx != null) {
            PackageManager pm = ctx.getPackageManager();
            ArrayList<Object> lst = PackageParserUtils.packageActivities(pkg);
            if (lst != null) {
                for (Object a: lst) {
                    PackageParserUtils.Activity aa = PackageParserUtils.Activity.fromComponent(a);
                    CompInfo info = new CompInfo();
                    info.component = aa;
                    info.fullPackageName = aa.getComponentName().getClassName();
                    info.enabled = pm.getComponentEnabledSetting(aa.getComponentName()) != PackageManager.COMPONENT_ENABLED_STATE_DISABLED;
                    lstComponentInfo.add(info);
                }
            }
        }
        return lstComponentInfo;
    }

    public static List<CompInfo> getServiceList(Context ctx, Object pkg) {
        List<CompInfo> lstComponentInfo = new ArrayList<>();
        if (ctx != null) {
            PackageManager pm = ctx.getPackageManager();
            ArrayList<Object> lst = PackageParserUtils.packageServices(pkg);
            if (lst != null) {
                for (Object a: lst) {
                    PackageParserUtils.Service aa = PackageParserUtils.Service.fromComponent(a);
                    CompInfo info = new CompInfo();
                    info.component = aa;
                    info.fullPackageName = aa.getComponentName().getClassName();
                    info.enabled = pm.getComponentEnabledSetting(aa.getComponentName()) != PackageManager.COMPONENT_ENABLED_STATE_DISABLED;
                    lstComponentInfo.add(info);
                }
            }
        }
        return lstComponentInfo;
    }

    public static List<CompInfo> getReceiverList(Context ctx, Object pkg) {
        List<CompInfo> lstComponentInfo = new ArrayList<>();
        if (ctx != null) {
            PackageManager pm = ctx.getPackageManager();
            ArrayList<Object> lst = PackageParserUtils.packageReceivers(pkg);
            if (lst != null) {
                for (Object a: lst) {
                    PackageParserUtils.Activity aa = PackageParserUtils.Activity.fromComponent(a);
                    CompInfo info = new CompInfo();
                    info.component = aa;
                    info.fullPackageName = aa.getComponentName().getClassName();
                    info.enabled = pm.getComponentEnabledSetting(aa.getComponentName()) != PackageManager.COMPONENT_ENABLED_STATE_DISABLED;
                    lstComponentInfo.add(info);
                }
            }
        }
        return lstComponentInfo;
    }

    public static List<CompInfo> getProviderList(Context ctx, Object pkg) {
        List<CompInfo> lstComponentInfo = new ArrayList<>();
        if (ctx != null) {
            PackageManager pm = ctx.getPackageManager();
            ArrayList<Object> lst = PackageParserUtils.packageProviders(pkg);
            if (lst != null) {
                for (Object a: lst) {
                    PackageParserUtils.Provider aa = PackageParserUtils.Provider.fromComponent(a);
                    CompInfo info = new CompInfo();
                    info.component = aa;
                    info.fullPackageName = aa.getComponentName().getClassName();
                    info.enabled = pm.getComponentEnabledSetting(aa.getComponentName()) != PackageManager.COMPONENT_ENABLED_STATE_DISABLED;
                    lstComponentInfo.add(info);
                }
            }
        }
        return lstComponentInfo;
    }

}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/utils/ComponentUtils.kt
@file:Suppress("Duplicates", "unused")

package com.rarnu.tools.neo.utils

import android.app.ActivityManager
import android.content.ComponentName
import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.util.Log

object ComponentUtils {

    @Suppress("DEPRECATION")
    fun isServiceRunning(context: Context, className: String?) = (context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager).getRunningServices(30).any { it.service.className == className }

    fun getActivityList(ctx: Context, pkg: PackageInfo) = mutableListOf<CompInfo>().apply {
        val pm = ctx.packageManager
        pkg.activities?.forEach {
            add(CompInfo().apply {
                Log.e("RootTools", "Activity => $it")
                pureName = it.name.substringAfterLast(".")
                componentClassName = it.name
                componentPackageName = it.packageName
                enabled = pm.getComponentEnabledSetting(ComponentName(it.packageName, it.name)) != PackageManager.COMPONENT_ENABLED_STATE_DISABLED
            })
        }
        sortBy { it.pureName }
    }.toList()

    fun getServiceList(ctx: Context, pkg: PackageInfo) = mutableListOf<CompInfo>().apply {
        val pm = ctx.packageManager
        pkg.services?.forEach {
            add(CompInfo().apply {
                pureName = it.name.substringAfterLast(".")
                componentClassName = it.name
                componentPackageName = it.packageName
                enabled = pm.getComponentEnabledSetting(ComponentName(it.packageName, it.name)) != PackageManager.COMPONENT_ENABLED_STATE_DISABLED
            })
        }
        sortBy { it.pureName }
    }.toList()

    fun getReceiverList(ctx: Context, pkg: PackageInfo) = mutableListOf<CompInfo>().apply {
        val pm = ctx.packageManager
        pkg.receivers?.forEach {
            add(CompInfo().apply {
                pureName = it.name.substringAfterLast(".")
                componentClassName = it.name
                componentPackageName = it.packageName
                enabled = pm.getComponentEnabledSetting(ComponentName(it.packageName, it.name)) != PackageManager.COMPONENT_ENABLED_STATE_DISABLED
            })
        }
        sortBy { it.pureName }
    }.toList()

    fun getProviderList(ctx: Context, pkg: PackageInfo) = mutableListOf<CompInfo>().apply {
        val pm = ctx.packageManager
        pkg.providers?.forEach {
            add(CompInfo().apply {
                pureName = it.name.substringAfterLast(".")
                componentClassName = it.name
                componentPackageName = it.packageName
                enabled = pm.getComponentEnabledSetting(ComponentName(it.packageName, it.name)) != PackageManager.COMPONENT_ENABLED_STATE_DISABLED
            })
        }
        sortBy { it.pureName }
    }.toList()
}

class CompInfo {

    var enabled = false
    var position = 0
    var pureName = ""
    var componentClassName: String? = null
    var componentPackageName: String? = null

    fun isServiceRunning(context: Context): Boolean {
        var ret = false
        try {
            ret = ComponentUtils.isServiceRunning(context, componentClassName)
        } catch (e: Throwable) {

        }
        return ret
    }

}